presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -321,6 +341,28 @@ public void fail()
         }
     }
 
+    @Override
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        noMorePagesForLifespan.add(lifespan);
+    }
+
+    @Override
+    public void registerLifespanFinishCallback(Consumer<Lifespan> callback)
+    {
+        checkState(lifespanFinishedCallback == null);
+        this.lifespanFinishedCallback = callback;
```

Comment:
IIUC a thread T1 can write to `lifespanFinishedCallback` here, and another thread T2 may be executing the page dereference handler on L238 and read `lifespanFinishedCallback`, which is a race.

Reply:
@nezihyigitbasi : My understanding is  lifespan callback is expected to be registered at the initialization of `SqlTaskExecution` , which should happen before any `enqueue` call. 

But I will double check on this.